### PR TITLE
Feature: Custom Functions

### PR DESCRIPTION
**Pull request by am11**
<del>Supersedes #644.</del>

### CODE REVIEW

File: src/binding.cpp [hunk 1]
```diff
@@ -233,117 +184,92 @@ void make_callback(uv_work_t* req) {
     node::FatalException(try_catch);
   }
 
-  if (ctx_w->importer_callback) {
-    uv_close((uv_handle_t*)&ctx_w->async, NULL);
-  }
+  //if (ctx_w->importer_bridge) {
+  //  uv_close((uv_handle_t*)&ctx_w->async, NULL);
+  //}
```
Comment (am11) [inline src/binding.cpp]: TODO: remove when everything else is fine.

**Comment by am11** [comment]
This is a historical moment!

historical-moment

File: src/create_string.cpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#include "create_string.h"
```
Comment (saper) [inline src/create_string.cpp]: Need

      #include <stdlib.h>

for `malloc()`

File: lib/index.js [hunk 3]
```diff
@@ -189,6 +188,27 @@ module.exports.render = function(options, cb) {
...
+
+      options.functions[signature] = function() {
+        var args = Array.prototype.slice.call(arguments),
+            bridge = args.pop();
+
+        function done(data) {
+          bridge.success(data);
+        }
+
+        var result = callback.apply(this, args.concat(done));
```
Comment (Jakobo) [inline lib/index.js]: `callback.apply(options.context, args.concat(done))` for consistency with importers?

Comment (am11) [inline lib/index.js]: Yes this is what I just told @matryo on gitter chat :)
This PR is experimental. the real meat is still #644. :wink:

Comment (Jakobo) [inline lib/index.js]: No worries. Let me know how (if) I can help. :)

Comment (matryo) [inline lib/index.js]: @Jakobo Yes this makes complete sense. Ultimately I can see that the context could be used to call built-in sass functions, or functions defined through the sass files.

Comment (Jakobo) [inline lib/index.js]: Totally. We've already seen the context object be useful in `import-once` importer development, and I have no doubt that image manipulation custom functions will memoize using the `context` for performance. :)

Comment (am11) [inline lib/index.js]: @Jakobo, wow image manipulation, like generating and mapping sprites? One day I will learn all the powers and mysteries behind `this.conext`, but I don't get time to actually *use* node-sass like others do! :smile_cat:

Comment (am11) [inline lib/index.js]: @Jakobo, now that #644 is merged, would you like to send a PR for `this.context` and the tests which goes with it? (-8

You may want to play with this cool new feature as well. :wink:

**Pull request closed without merging**